### ISSUE TITLE
Epic: Riemann deep-specialization branch

### ISSUE THREAD

**Issue by murillo128** [murillo128/qwen-lean#36]
# Riemann deep-specialization branch

This epic tracks an **experimental specialist branch** of qwen-lean focused on prime number theory and the mathematical neighborhood of the Riemann zeta function / Riemann Hypothesis.

It is deliberately separate from the generic `qwen-lean` roadmap in #1. The goal is not to improve the generic model by contaminating it with a narrow target distribution; the goal is to measure how far aggressive domain specialization can push a trainable open-weight Lean model on one tightly connected mathematical region.

## Core question

How much can a relatively small Lean prover gain from a deliberately narrow, high-density curriculum around one mathematical research area, and how does that compare with a generic Lean model and stronger generalist/specialist foundations?

The experiment should distinguish:

- generic Lean capability;
- broad number-theory specialization;
- local Riemann/zeta specialization;
- extreme repeated/overfit specialization;
- literal target leakage/memorization controls.

The last category is allowed only as an explicitly labeled control. It must never be mixed into clean generalization results.

## Timing / dependency

**Data work can start now. Training cannot.**

The theorem inventory, relevance graph and specialist corpora are foundation-model-independent and may be built immediately from the already accepted Phase 2 mathlib corpus and pinned source identity. In parallel, external Lean projects, other formal systems and mathematical literature can be inventoried to build a broader Riemann theorem atlas without treating unverified material as training data.

Actual specialist training waits until the cross-model assessment in #27 is sufficiently complete to choose the foundation model. Do not assume the specialist must descend from the current Qwen3-8B `reference-sft-v1`; a Qwen3.5/3.6 or another trainable open-weight model may be a better starting point.

The specialist model is a separate artifact/family from generic `qwen-lean`.

## Current execution

- [ ] #37 — build the theorem relevance graph and materialize `number-theory-wide-v1`, `riemann-bubble-v1`, `riemann-core-v1` plus clean specialist holdouts from the accepted Phase 2/mathlib corpus. **Execution-ready; no GPU/model-selection dependency.**
- [ ] #38 — build a source-bounded external Riemann theorem atlas covering external Lean formalizations, other proof assistants, and literature-level equivalences/implications/partial results. **Execution-ready; no GPU/model-selection dependency.**

Later training/RL children should be created only after #37 tells us what verified internal data exists, #38 maps useful external/formalization opportunities, and #27 informs foundation-model choice.

## Intended corpus structure

Do not create one undifferentiated `grep prime` dataset. Build a graph/bubble with auditable provenance and multiple rings, for example:

1. **seed/core** — declarations explicitly about primes, Riemann zeta, Dirichlet/L-series, prime-counting/asymptotic statements, and directly relevant analytic-number-theory objects;
2. **direct-premise ring** — declarations used by or directly using those seeds, using resolved premise metadata where available;
3. **multi-hop dependency ring** — bounded graph expansion over premise/dependency edges;
4. **structural/source-neighborhood ring** — nearby declarations/modules needed to preserve coherent local mathematical context;
5. **broader number-theory control ring** — related number theory outside the immediate Riemann bubble.

Every retained record in the first specialist corpora must map back to verified source identity and proof. External theorem-atlas entries are not training records merely because they are mathematically relevant.

## Clean evaluation boundary

Before specialist training, freeze evaluation partitions that are not used for optimizer updates or training-time model selection.

At minimum preserve distinct notions of:

- **near holdout** — unseen theorems very close to the Riemann bubble;
- **far/domain holdout** — unseen number-theory theorems outside the immediate bubble;
- **generic Lean control** — existing project workloads to measure catastrophic forgetting;
- **memorization/leakage control** — deliberately contaminated examples, if used, reported separately.

Random theorem-level splitting is not sufficient when dependency-neighbor theorems are near-duplicates in mathematical content. Prefer graph/source/component boundaries and explicit statement/dependency overlap checks.

The Riemann Hypothesis itself may be represented as a target statement/research anchor, but no result from this project should be described as a proof of RH unless Lean verifies an actual proof without placeholders and the mathematics is independently scrutinized. The useful scientific signal is expected to come primarily from related provable holdouts and transfer within the specialist domain.

## Later specialist training family

After #27 informs foundation-model choice, the likely experimental family is:

```text
chosen open-weight foundation
        |
        +--> generic Lean path / control
        |
        +--> Riemann specialist SFT
                  |
                  +--> specialist RL / verifier-reward branch
                  |
                  +--> extreme-overfit / memorization-control branch
```

Do not force the generic and specialist models to share the same final ancestry if evidence favors a newer foundation for the specialist.

External Lean material discovered by #38 should enter a later corpus only through an explicit ingestion/formalization issue with pinned provenance, license/toolchain checks and Lean verification.

## Comparisons we ultimately want

Report the specialist against at least:

- the chosen unadapted foundation;
- a generic Lean SFT/control at comparable scale where available;
- the specialist SFT checkpoint;
- specialist RL if run;
- any extreme-overfit/memorization-control checkpoint separately.

Key outcomes include Lean pass@k on clean near/far holdouts, train-vs-holdout gap, exact-target reproduction where meaningful, generated-proof behavior, and loss of generic Lean capability.

The most interesting positive result is **local generalization**: solving unseen Riemann/number-theory-near theorems rather than merely reproducing retained training proofs.

## Non-goals / invariants

- Do not modify the generic qwen-lean training corpus to make this experiment look better.
- Do not contaminate existing miniF2F or other retained evaluation controls.
- Do not choose the final foundation model before the model assessment evidence warrants it.
- Do not treat literal training-set reproduction as evidence of mathematical generalization.
- Do not treat literature statements or proofs in another proof assistant as Lean-verified training data.
- Keep full generated corpora/model checkpoints outside Git; commit code/config/manifests/statistics/compact evidence.
- Preserve Lean verification as the proof-success oracle.

## Relationship to other work

- #1 remains the generic qwen-lean roadmap.
- #27 owns the current cross-model assessment and is the main model-selection input for this branch.
- #37 owns the immediate verified mathlib-derived theorem/corpus work.
- #38 owns the broader external theorem/formalization/literature atlas.
- The accepted Phase 2 corpus/extraction machinery is the default source for the initial specialist-data work.

This epic may later gain separate external-Lean-ingestion, Lean-formalization, training and RL children after #37/#38 statistics and #27 results make those designs concrete. Avoid prematurely freezing training hyperparameters or a specific Qwen generation here.
